torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -1882,6 +1885,10 @@ void UpdateReliable(
         output->node()->kind().toDisplayString(),
         " type is missing, so it may result in wrong shape inference for the exported graph. ",
         "Please consider adding it in symbolic function.");
+    diagnostics::Diagnose(
```

Comment:
> embed `TORCH_WARN`

Yes the actual emission can reuse whatever previous functionalities, potentially configurable too with DiagnosticOptions. Note that cpp api is supposed to be a thin wrapper over the actual library in Python, where the emission happens.

> optional `user_log`

This can be achieved with the `additional_message` argument. A bit more detail on how these info can be supplied. I created `diagostic.with_additional_message()` method to enable contiguous build up, but it makes it hard to decide emission time. Perhaps we'd still want a single constructor like function to record a diagnostic with all the info as arguments and be done with it? Let me know your thoughts.

> maps to a friendly message that is printed to stdout

Each rule does have several messages defined with them, they expose different level of details and information. The most informative one will also be used for generating the doc page of that rule under torch.onnx doc. You can view it at `rules.yaml`. (Will add proper documentation for these ... :))

Reply:
IIUC, `diagnostics::Diagnose` **already** calls `TORCH_WARN` through the use of `user_log` and `additional_message` and lines 1883-1887 could be removed?

Or that would be the behavior in a follow up PR?

Reply:
Right now `Diagnose` doesn't have any real user interacting behavior, i.e. it's not printing anything. Following up the emission of diagnosing will be added. When that happens, TORCH_WARN won't be needed. I'm adding comments above this to mention it's experimental and does not print anything yet.